openprescribing/frontend/management/commands/scrape_numbers.py

```diff
@@ -0,0 +1,243 @@
+    for k, v in get_resolver().reverse_dict.items():
+
+        # Ignore records where the key is a view function.
+        if not isinstance(k, str):
+            continue
+
+        name = k
+        pattern = v[0][0][0]
+        keys = v[0][0][1]
+
+        # Ignore records starting with prefixes we're not interested in.
+        if any(pattern.startswith(prefix) for prefix in prefixes_to_ignore):
+            continue
+
+        # Ignore any URLs that 
+        if '%' not in pattern and 'all-england' not in pattern:
+            continue
+
+        path = bulid_path(pattern, keys)
+        yield name, path
+
+
+def bulid_path(pattern, keys):
+    '''Replace placeholders in `pattern` to build path to be scraped.
+
+    `pattern` will be a string like "measure/%(measure)s/ccg/%(entity_code)s/".
+    `keys` will be a list like ["measure", "entity_code"].
+    '''
+
+    substitutions = {
+        'practice_code': 'L83100',
+        'ccg_code': '15N',
+        'stp_code': 'E54000037',
+        'regional_team_code': 'Y58',
+        'bnf_code': '0205051R0BBAIAN',
+        'measure': 'ace',
+    }
+
+    path = pattern
+
+    for key in keys:
+        subst_key = None
+
+        if key in ['code', 'entity_code']:
+            for token in ['practice', 'ccg', 'stp', 'regional-team']:
+                if token in pattern:
+                    subst_key = '{}_code'.format(token).replace('-', '_')
+            if subst_key is None:
+                subst_key = 'bnf_code'
+        else:
+            subst_key = key
+
+        path = path.replace(
+            '%({})s'.format(key), 
+            substitutions[subst_key]
+        )
+
+    assert '%' not in path, 'Could not interpolate ' + name
+
+    return path
+
+
+def extract_numbers(browser, path, name, log_path):
+    '''Request URL, write copy of response to log_path, and return list of
+    anything that looks like an interesting number in the response.
+    '''
+
+    url = 'https://openprescribing.net/' + path
+    browser.get(url)
+
+    # Wait until all AJAX requests are complete.
+    while browser.execute_script('return jQuery.active > 0'):
+        sleep(0.1)
+
+    source = browser.page_source
+
+    with open(os.path.join(log_path, name + '.html'), 'w') as f:
+        f.write(source.encode('utf8'))
+
+    doc = BeautifulSoup(source, 'html.parser')
+    body = doc.find('body')
+
+    # The SVG charts contain all sorts of things that look like numbers (eg
+    # rgb(255,255,255)) so we drop all <svg> elements.
+    for tag in body.find_all('svg'):
+        tag.decompose()
+
+    rx = re.compile(u'''
+        £[\d,\.]+           # Anything that looks like a cost
+        |                   # ...or...
+        \d{1,3}(?:,\d{3})+  # anything that looks like a humanized number
+        ''',
+        re.VERBOSE
+    )
+    return rx.findall(body.text)
+
+
+def compare_numbers(previous_numbers, numbers):
+    '''Compare dictionaries of numbers, returning list of any differences.
+    '''
+
```

Comment:
Just a thought: using a unified diff format for comparing the lists of numbers (using `difflib`) might make it easier to see what's going on.  (Or even for comparing the whole logs, if they format well as JSON).

Reply:
Oh yes, now I think about it I used that for diffing JSON during my matrixstore smoketesting:
```python
    ...
    diff = difflib.unified_diff(as_text_lines(item_a), as_text_lines(item_b), n=8)
    return '\n'.join(diff)

def as_text_lines(data):
    return json.dumps(data, indent=2, sort_keys=True).splitlines()
```

Reply:
Nice idea, but I don't think it'd be particularly helpful, as we don't have information about where any of the numbers are from.  I anticipate that if we find that any numbers have changed over the deploy, we'll need to visually compare the old and new versions of the pages (which we've saved) to understand what the difference is.